src/x11/x.c

```diff
@@ -110,10 +110,13 @@ static void x_win_round_corners(struct window_x11 *win, const int rad)
 
         /* To mark all pixels, which should get exposed, we
          * use a circle for every corner and two overlapping rectangles */
+        int fw = settings.frame_width / 2 + settings.frame_width % 2;
+        if (fw < 1)
+                fw = 1;
         unsigned const int centercoords[] = {
-                0,               0,
-                width - dia - 1, 0,
-                0,               height - dia - 1,
+                -fw,             -fw,
+                width - dia - 1, -fw,
+                -fw,             height - dia - 1,
```

Comment:
> Yet can't understand, where 0.5,0.5 displacement is set (originall integer coordinates would point to pixel corner instead of its center).

This was bothering me since the beginning and I finally figured out why: According to the [X11 docs](https://tronche.com/gui/x/xlib/graphics/filling-areas/XFillArc.html), the coordinates in XFillArc "specify the upper-left corner of the bounding rectangle", so at (0,0) this includes the frame, this change over-compensates and moves the arcs too much, breaking the round corners if the frame is large enough.

I'd only put 1 or 2 pixels of leeway here. The transparency changes are awesome, but they only work if there's a running compositor. Limiting the number of transparent pixels preserves the AA effect but minimizes the black artifacts in the corners.

Reply:
It's not center, XDrawArc takes rectangle coords, containing arc. Did not test all frame width, only 1 and 0. Imho not sure, what's better - aliased border or several too dark pixels. Comparing with Xfce4-notifyd it just disabled rounding for no-comp.